oci/pull.bzl

```diff
@@ -0,0 +1,156 @@
+
+    if resource != "blobs" and resource != "manifests":
+        fail("resource must be blobs or manifests")
+
+    # Construct the URL to fetch from remote, see
+    # https://github.com/google/go-containerregistry/blob/62f183e54939eabb8e80ad3dbc787d7e68e68a43/pkg/v1/remote/descriptor.go#L234
+    firstslash = rctx.attr.image.find("/")
+    registry_url = "https://{registry}/v2/{repository}/{resource}/{identifier}".format(
+        registry = rctx.attr.image[:firstslash],
+        repository = rctx.attr.image[firstslash + 1:],
+        resource = resource,
+        identifier = identifier,
+    )
+
+    sha256 = None
+    if identifier.startswith("sha256:"):
+        sha256 = identifier[len("sha256:"):]
+    else:
+        # buildifier: disable=print
+        print("WARNING: fetching from {} without a digest, check your oci_pull configuration")
+    rctx.download(
+        output = output,
+        sha256 = sha256,
+        url = registry_url,
+    )
+    if resource == "manifests":
+        return json.decode(rctx.read(output))
+    return None
+
+_build_file = """\
+"Generated by oci_pull"
+
+load("@aspect_bazel_lib//lib:copy_to_directory.bzl", "copy_to_directory")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+
+package(default_visibility = ["//visibility:public"])
+
+# Mimic the output of crane pull [image] layout --format=oci
+write_file(
+    name = "write_layout",
+    out = "oci-layout",
+    content = [json.encode({{
+        "imageLayoutVersion": "1.0.0"
+    }})],
+)
+
+# TODO: should we transform this to include only manifests with matching platform?
+# We copy the entire index with all platforms, while crane writes this file with only the one platform.
+write_file(
+    name = "write_index",
+    out = "index.json",
+    content = [\"\"\"{index_content}\"\"\"],
+)
+
+copy_to_directory(
+    name = "blobs",
+    out = "blobs/sha256",
+    include_external_repositories = ["*"],
+    srcs = {tars} + [
+        ":{manifest_file}",
+        ":{config_file}",
+    ],
+)
+
+copy_to_directory(
+    name = "image",
+    out = "layout",
+    include_external_repositories = ["*"],
+    srcs = [
+        "blobs",
+        "oci-layout",
+        "index.json",
+    ],
+)
+"""
+
+_alias_target = """\
+alias(
+    name = "{name}",
+    actual = select({{
+        "@platforms//cpu:arm64": "@{name}_linux_arm64//:image",
+        "@platforms//cpu:x86_64": "@{name}_linux_amd64//:image",
+    }}),
+    visibility = ["//visibility:public"],
+)
+"""
+
+def _pull_impl(rctx):
+    image_mf_file = rctx.attr.reference.replace("sha256:", "")
+    image_mf = _download(rctx, rctx.attr.reference, image_mf_file, resource = "manifests")
+
+    image_config_file = image_mf["config"]["digest"].replace("sha256:", "")
+    image_config = _download(rctx, image_mf["config"]["digest"], image_config_file)
+    tars = []
+    for layer in image_mf["layers"]:
+        sha256 = layer["digest"].replace("sha256:", "")
```

Comment:
nit: while `sha256` is a common hash algorithm, the spec allows other algorithms as well. like `sha512` and blake. for MVP we could keep it like this.

Reply:
I extracted a function for this trimming.
Also added TODOs anywhere I have sha256 specifically, pointing to an issue